### PR TITLE
Exception handling in control flow

### PR DESCRIPTION
**Pull request by yadavcbala**
*Issue #90*

*Description of changes:*

The description of the changes is moved to Issue#90.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: situp-core/src/main/java/com/amazon/situp/Situp.java [hunk 1]
```diff
@@ -56,25 +56,25 @@ public boolean execute(final String configurationFileLocation) {
         LOG.info("Using {} configuration file", configurationFileLocation);
         final PipelineParser pipelineParser = new PipelineParser(configurationFileLocation);
         transformationPipelines = pipelineParser.parseConfiguration();
+        if (transformationPipelines.size() == 0){
+            LOG.warn("No valid pipeline is available for execution, exiting");
```
Comment (laneholloway) [inline situp-core/src/main/java/com/amazon/situp/Situp.java:60]: Isn't this an error and not a warning?

File: situp-core/src/main/java/com/amazon/situp/pipeline/PipelineConnector.java [hunk 2]
```diff
@@ -34,6 +34,7 @@ public void start(final Buffer<T> buffer) {
 
     @Override
     public void stop() {
+        //this.buffer = null;
```
Comment (laneholloway) [inline situp-core/src/main/java/com/amazon/situp/pipeline/PipelineConnector.java:37]: Do you need this?

Comment (yadavcbala) [reply situp-core/src/main/java/com/amazon/situp/pipeline/PipelineConnector.java:37]: good catch

File: situp-core/src/main/java/com/amazon/situp/Situp.java (hunk 1 above)
Comment (yadavcbala) [reply situp-core/src/main/java/com/amazon/situp/Situp.java:60]: its an error

File: situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java [hunk 3]
```diff
@@ -113,9 +126,12 @@ private int getDefaultProcessorThreads() {
             final PluginSetting pluginSetting,
             final Map<String, Pipeline> pipelineMap,
             final Map<String, PipelineConfiguration> pipelineConfigurationMap) {
+        LOG.info("Building [{}] as source component for the pipeline [{}]", pluginSetting.getName(), sourcePipelineName);
         final Optional<String> pipelineNameOptional = getPipelineNameIfPipelineType(pluginSetting);
```
Comment (kowshikn) [inline situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java:130]: Pipeline.Name has to be present we are using PipelineConnector right? Given that we handle the exceptions, we can now avoid the Optional and throw exception.

File: situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java [hunk 4]
```diff
@@ -125,6 +141,7 @@ private int getDefaultProcessorThreads() {
     }
 
     private Sink buildSinkOrConnector(final PluginSetting pluginSetting) {
+        LOG.info("Building [{}] as sink component", pluginSetting.getName());
         final Optional<String> pipelineNameOptional = getPipelineNameIfPipelineType(pluginSetting);
         if (pipelineNameOptional.isPresent()) { //update to ifPresentOrElse when using JDK9
```
Comment (kowshikn) [inline situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java:146]: Same as above.

File: situp-core/src/main/java/com/amazon/situp/pipeline/Pipeline.java [hunk 5]
```diff
@@ -146,43 +145,49 @@ public int getReadBatchTimeoutInMillis() {
...
+        } catch (Exception ex) {
+            //source failed to start - Cannot proceed further with the current pipeline
+            LOG.error("Pipeline [{}] encountered exception while starting the source, skipping execution", name, ex);
+            //skip further processing of this pipeline
+            return;
+        }
+        LOG.info("Pipeline [{}] - Submitting request to initiate the pipeline processing", name);
+        for (int i = 0; i < processorThreads; i++) {
+            processorSinkExecutorService.submit(new ProcessWorker(buffer, processors, sinks, this));
+        }
```
Comment (kowshikn) [inline situp-core/src/main/java/com/amazon/situp/pipeline/Pipeline.java:160]: nit: can this not be inside the `try` block? you can avoid the `return`

File: situp-core/src/main/java/com/amazon/situp/pipeline/common/PipelineThreadPoolExecutor.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    private final Pipeline pipeline;
+
+    public PipelineThreadPoolExecutor(
+            final int corePoolSize,
+            final int maximumPoolSize,
+            final long keepAliveTime,
+            final TimeUnit unit,
+            final BlockingQueue<Runnable> workQueue,
+            final ThreadFactory threadFactory,
+            final Pipeline pipeline) {
```
Comment (kowshikn) [inline situp-core/src/main/java/com/amazon/situp/pipeline/common/PipelineThreadPoolExecutor.java:27]: nit: In the processworker you are dependent on lot other members of the pipeline, here you don't have any thing except pipeline::Name, can you not accept a string.

File: situp-core/src/main/java/com/amazon/situp/pipeline/common/PipelineThreadPoolExecutor.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                        "cannot proceed further", pipeline.getName(), ex);
+                pipeline.shutdown();
+            } catch (InterruptedException ex) {
+                LOG.error("Pipeline [{}] process worker encountered a fatal exception, terminating", pipeline.getName(), ex);
+                pipeline.shutdown();
+                Thread.currentThread().interrupt();
+            }
+        }
+        // If we ever use the execute instead of submit
+        if (throwable != null) {
```
Comment (kowshikn) [inline situp-core/src/main/java/com/amazon/situp/pipeline/common/PipelineThreadPoolExecutor.java:76]: nit: else if

File: situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java (hunk 3 above)
Comment (yadavcbala) [reply situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java:130]: if the configured source is of type pipeline, then it will return pipeline name; if it is not of pipeline type, then we won't have name. so i think we need Optional.

File: situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java (hunk 4 above)
Comment (yadavcbala) [reply situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java:146]: if the configured sink is of type pipeline, then it will return pipeline name; if it is not of pipeline type, then we won't have name. so i think we need Optional.

File: situp-core/src/main/java/com/amazon/situp/pipeline/common/PipelineThreadPoolExecutor.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class PipelineThreadPoolExecutor extends ThreadPoolExecutor {
```
Comment (kowshikn) [inline situp-core/src/main/java/com/amazon/situp/pipeline/common/PipelineThreadPoolExecutor.java:16]: nit: This class could have a java doc explaining the reason why we need this.

File: situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java (hunk 3 above)
Comment (kowshikn) [reply situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java:130]: My bad, this makes sense.

**Review by kowshikn — APPROVED** [review]

File: situp-core/src/main/java/com/amazon/situp/pipeline/common/PipelineThreadPoolExecutor.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The {@link Pipeline} process workers which execute pipeline processors and sinks are spun in n different threads
```
Comment (laneholloway) [inline situp-core/src/main/java/com/amazon/situp/pipeline/common/PipelineThreadPoolExecutor.java:17]: `in n` looks like a typo, you might want to capitalize the `N` or reword it to make it more clear.. perhaps something like this?
`The Pipeline process workers that execute pipeline processors and sinks each run in independent threads. If any of these threads fail, have an error, or throw an exception, the pipeline needs to shutdown. We extend the ThreadPoolExecutor to ....`

**Review by laneholloway — APPROVED** [review]
